File: apps/api/src/investhome_api/schemas/crm_leads.py
"""CRM operational lead workspace schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

CrmLeadStage = Literal["yeni", "contacted", "following", "qualified", "converted", "unqualified"]
CrmLeadIngestStatus = Literal["ok", "unmatched", "failed"]


class CrmLeadMatch(BaseModel):
    kind: Literal["lead", "person"]
    id: UUID
    name: str
    phone: str | None = None
    email: str | None = None
    reason: str
    href: str | None = None


class CrmLeadOwnerOption(BaseModel):
    id: UUID
    name: str


class CrmLeadActivityItem(BaseModel):
    id: UUID
    action: str
    description: str
    actor_name: str | None = None
    created_at: datetime
    metadata: dict[str, Any] | None = None


class CrmLeadItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    full_name: str
    phone: str | None = None
    email: str | None = None
    source: str | None = None
    campaign: str | None = None
    ad_id: str | None = None
    form_id: str | None = None
    project: str | None = None
    owner_user_id: UUID | None = None
    owner_name: str | None = None
    stage: CrmLeadStage
    notes: str | None = None
    provider: str | None = None
    ingest_status: str
    converted_contact_id: UUID | None = None
    converted_contact_name: str | None = None
    created_at: datetime
    updated_at: datetime


class CrmLeadDetail(CrmLeadItem):
    existing_person_id: UUID | None = None
    existing_person_name: str | None = None
    metadata: dict[str, Any] | None = None
    activity: list[CrmLeadActivityItem] = Field(default_factory=list)


class CrmLeadKpis(BaseModel):
    total: int = 0
    active: int = 0
    yeni: int = 0
    following: int = 0
    qualified: int = 0
    converted: int = 0
    unqualified: int = 0
    unmatched: int = 0
    failed: int = 0


class CrmLeadListResponse(BaseModel):
    items: list[CrmLeadItem]
    total: int
    kpis: CrmLeadKpis
    sources: list[str] = Field(default_factory=list)
    projects: list[str] = Field(default_factory=list)
    owners: list[CrmLeadOwnerOption] = Field(default_factory=list)
    stages: list[str] = Field(
        default_factory=lambda: ["yeni", "contacted", "following", "qualified", "converted", "unqualified"]
    )


class CrmLeadWrite(BaseModel):
    full_name: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=50)
    email: str | None = Field(default=None, max_length=255)
    source: str | None = Field(default=None, max_length=100)
    campaign: str | None = Field(default=None, max_length=255)
    ad_id: str | None = Field(default=None, max_length=120)
    form_id: str | None = Field(default=None, max_length=120)
    project: str | None = Field(default=None, max_length=255)
    owner_user_id: UUID | None = None
    notes: str | None = Field(default=None, max_length=5000)
    stage: CrmLeadStage | None = None


class CrmLeadCreate(CrmLeadWrite):
    provider: str | None = Field(default="manual", max_length=40)


class CrmLeadUpdate(CrmLeadWrite):
    pass


class CrmLeadStageUpdate(BaseModel):
    stage: CrmLeadStage


class CrmLeadConvertRequest(BaseModel):
    confirm_existing_person_id: UUID | None = None


class CrmLeadConvertResponse(BaseModel):
    lead: CrmLeadDetail
    contact_id: UUID
    contact_name: str
    reused_existing: bool
    href: str


class CrmLeadIngestRequest(BaseModel):
    provider: str = Field(min_length=1, max_length=40)
    source: str | None = Field(default=None, max_length=100)
    campaign: str | None = Field(default=None, max_length=255)
    ad_id: str | None = Field(default=None, max_length=120)
    form_id: str | None = Field(default=None, max_length=120)
    full_name: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=50)
    email: str | None = Field(default=None, max_length=255)
    project: str | None = Field(default=None, max_length=255)
    notes: str | None = Field(default=None, max_length=5000)
    payload: dict[str, Any] | None = None


class CrmLeadConflictBody(BaseModel):
    code: Literal["existing_lead", "existing_person", "ambiguous_person"]
    message: str
    matches: list[CrmLeadMatch] = Field(default_factory=list)

File: apps/web/src/workspaces/crm/api/crm-leads.ts
import { apiFetch, ApiError, getApiBaseUrl } from '@/lib/api/client';

export type CrmLeadStage =
  | 'yeni'
  | 'contacted'
  | 'following'
  | 'qualified'
  | 'converted'
  | 'unqualified';

export type CrmLeadMatch = {
  kind: 'lead' | 'person';
  id: string;
  name: string;
  phone?: string | null;
  email?: string | null;
  reason: string;
  href?: string | null;
};

export type CrmLeadOwnerOption = {
  id: string;
  name: string;
};

export type CrmLeadActivityItem = {
  id: string;
  action: string;
  description: string;
  actor_name?: string | null;
  created_at: string;
  metadata?: Record<string, unknown> | null;
};

export type CrmLeadItem = {
  id: string;
  full_name: string;
  phone: string | null;
  email: string | null;
  source: string | null;
  campaign: string | null;
  ad_id: string | null;
  form_id: string | null;
  project: string | null;
  owner_user_id: string | null;
  owner_name: string | null;
  stage: CrmLeadStage;
  notes: string | null;
  provider: string | null;
  ingest_status: string;
  converted_contact_id: string | null;
  converted_contact_name: string | null;
  created_at: string;
  updated_at: string;
  existing_person_id?: string | null;
  existing_person_name?: string | null;
  activity?: CrmLeadActivityItem[];
};

export type CrmLeadKpis = {
  total: number;
  active: number;
  yeni: number;
  following: number;
  qualified: number;
  converted: number;
  unqualified: number;
  unmatched: number;
  failed: number;
};

export type CrmLeadListResponse = {
  items: CrmLeadItem[];
  total: number;
  kpis: CrmLeadKpis;
  sources: string[];
  projects: string[];
  owners: CrmLeadOwnerOption[];
  stages: CrmLeadStage[];
};

export type CrmLeadWrite = {
  full_name?: string;
  phone?: string;
  email?: string;
  source?: string;
  campaign?: string;
  ad_id?: string;
  form_id?: string;
  project?: string;
  owner_user_id?: string | null;
  notes?: string;
  stage?: CrmLeadStage;
};

export type CrmLeadConvertResponse = {
  lead: CrmLeadItem;
  contact_id: string;
  contact_name: string;
  reused_existing: boolean;
  href: string;
};

export type CrmLeadListParams = {
  search?: string;
  stage?: string;
  source?: string;
  project?: string;
  owner_id?: string;
  date_from?: string;
  date_to?: string;
  ingest_status?: string;
};

export class CrmLeadConflictError extends Error {
  readonly status = 409;
  readonly code: string;
  readonly matches: CrmLeadMatch[];

  constructor(code: string, message: string, matches: CrmLeadMatch[]) {
    super(message);
    this.name = 'CrmLeadConflictError';
    this.code = code;
    this.matches = matches;
  }
}

function buildParams(values: Record<string, string | undefined>): URLSearchParams {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(values)) {
    if (!value) continue;
    params.set(key, value);
  }
  return params;
}

export async function fetchCrmLeads(params: CrmLeadListParams = {}): Promise<CrmLeadListResponse> {
  const query = buildParams({
    search: params.search,
    stage: params.stage,
    source: params.source,
    project: params.project,
    owner_id: params.owner_id,
    date_from: params.date_from,
    date_to: params.date_to,
    ingest_status: params.ingest_status,
  });
  const suffix = query.toString() ? `?${query.toString()}` : '';
  return apiFetch<CrmLeadListResponse>(`/crm/leads${suffix}`);
}

export async function fetchCrmLead(id: string): Promise<CrmLeadItem> {
  return apiFetch<CrmLeadItem>(`/crm/leads/${id}`);
}

async function parseConflict(response: Response, fallback: string): Promise<never> {
  const body = (await response.json().catch(() => ({}))) as {
    detail?: string;
    error?: { code?: string; message?: string };
    matches?: CrmLeadMatch[];
  };
  throw new CrmLeadConflictError(
    body.error?.code || 'existing_person',
    body.error?.message || body.detail || fallback,
    body.matches ?? [],
  );
}

export async function createCrmLead(payload: CrmLeadWrite, confirm = false): Promise<CrmLeadItem> {
  const response = await fetch(`${getApiBaseUrl()}/crm/leads${confirm ? '?confirm=true' : ''}`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    cache: 'no-store',
    body: JSON.stringify(payload),
  });
  if (response.status === 409) {
    await parseConflict(response, 'Mevcut kayıt bulundu');
  }
  if (!response.ok) {
    throw new ApiError(`Request failed with status ${response.status}`, response.status);
  }
  return response.json() as Promise<CrmLeadItem>;
}

export async function updateCrmLead(id: string, payload: CrmLeadWrite): Promise<CrmLeadItem> {
  return apiFetch<CrmLeadItem>(`/crm/leads/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
}

export async function moveCrmLeadStage(id: string, stage: CrmLeadStage): Promise<CrmLeadItem> {
  return apiFetch<CrmLeadItem>(`/crm/leads/${id}/stage`, {
    method: 'POST',
    body: JSON.stringify({ stage }),
  });
}

export async function convertCrmLead(
  id: string,
  confirmExistingPersonId?: string,
): Promise<CrmLeadConvertResponse> {
  const response = await fetch(`${getApiBaseUrl()}/crm/leads/${id}/convert`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    cache: 'no-store',
    body: JSON.stringify({
      confirm_existing_person_id: confirmExistingPersonId || null,
    }),
  });
  if (response.status === 409) {
    await parseConflict(response, 'Kişi eşleşmesi belirsiz');
  }
  if (!response.ok) {
    throw new ApiError(`Request failed with status ${response.status}`, response.status);
  }
  return response.json() as Promise<CrmLeadConvertResponse>;
}
